ostd/src/mm/dma/dma_coherent.rs

```diff
@@ -1,186 +1,177 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use core::ops::Deref;
+use core::{fmt::Debug, mem::ManuallyDrop};
 
-use cfg_if::cfg_if;
-
-use super::{DmaError, check_and_insert_dma_mapping, remove_dma_mapping};
+use super::util::{alloc_kva, dma_remap, has_tdx, split_daddr, unmap_dma_remap};
+#[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+use super::util::{alloc_unprotect_physical_range, dealloc_protect_physical_range};
 use crate::{
-    arch::iommu,
+    error::Error,
     mm::{
-        HasDaddr, HasPaddr, HasSize, Infallible, PAGE_SIZE, Paddr, USegment, VmReader, VmWriter,
-        dma::{Daddr, DmaType, dma_type},
+        Daddr, FrameAllocOptions, HasDaddr, HasPaddr, HasPaddrRange, HasSize, Infallible,
+        PAGE_SIZE, Paddr, Segment, Split, VmReader, VmWriter,
         io_util::{HasVmReaderWriter, VmReaderWriterIdentity},
-        kspace::{KERNEL_PAGE_TABLE, paddr_to_vaddr},
-        page_prop::CachePolicy,
+        kspace::kvirt_area::KVirtArea,
     },
 };
 
-cfg_if! {
-    if #[cfg(all(target_arch = "x86_64", feature = "cvm_guest"))] {
-        use crate::arch::tdx_guest;
-    }
-}
-
-/// A coherent (or consistent) DMA mapping,
-/// which guarantees that the device and the CPU can
-/// access the data in parallel.
-///
-/// The mapping will be destroyed automatically when
-/// the object is dropped.
+/// A DMA memory object with coherent cache.
 #[derive(Debug)]
 pub struct DmaCoherent {
-    segment: USegment,
-    start_daddr: Daddr,
+    inner: Inner,
+    map_daddr: Option<Daddr>,
     is_cache_coherent: bool,
 }
 
+#[derive(Debug)]
+enum Inner {
+    Segment(Segment<()>),
+    Kva(KVirtArea, Paddr),
+}
+
 impl DmaCoherent {
-    /// Creates a coherent DMA mapping backed by `segment`.
-    ///
-    /// The `is_cache_coherent` argument specifies whether
-    /// the target device that the DMA mapping is prepared for
-    /// can access the main memory in a CPU cache coherent way
-    /// or not.
+    /// Allocates a region of physical memory for coherent DMA access.
     ///
-    /// The method fails if any part of the given `segment`
-    /// already belongs to a DMA mapping.
-    pub fn map(segment: USegment, is_cache_coherent: bool) -> core::result::Result<Self, DmaError> {
-        let paddr = segment.paddr();
-        let frame_count = segment.size() / PAGE_SIZE;
-
-        if !check_and_insert_dma_mapping(paddr, frame_count) {
-            return Err(DmaError::AlreadyMapped);
-        }
+    /// If the device can access the memory with coherent access to the CPU
+    /// cache, set `is_cache_coherent` to `true`.
+    pub fn alloc(nframes: usize, is_cache_coherent: bool) -> Result<Self, Error> {
+        let has_tdx = has_tdx();
```

Comment:
I really don't like to see arch-specific concept in arch-agnostic module. This forces people who are not interested in TDX (e.g., a RISC-V developer) to learn and care about TDX. I think a much better name would be `is_inside_cvm`, which tells whether the kernel is running inside a CVM (confidential VMs). In all CVMs, guest physical memory are secure by default.